core/cmd/xdev/internal/cmd/build.go

```diff
@@ -0,0 +1,310 @@
+package cmd
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"github.com/xuperchain/xuperchain/core/cmd/xdev/internal/mkfile"
+)
+
+var (
+	defaultCxxFlags = []string{
+		"-std=c++11",
+		"-Os",
+		"-I/usr/local/include",
+		"-Isrc",
+		"-Werror=vla",
+	}
+	defaultLDFlags = []string{
+		"-Oz",
+		"-s TOTAL_STACK=256KB",
+		"-s TOTAL_MEMORY=1MB",
+		"-s DETERMINISTIC=1",
+		"-s EXTRA_EXPORTED_RUNTIME_METHODS=[\"stackAlloc\"]",
+		"-L/usr/local/lib",
+		"-lprotobuf-lite",
+		"-lpthread",
+	}
+)
+
+type buildCommand struct {
+	cxxFlags []string
+	ldflags  []string
+	builder  *mkfile.Builder
+	entryPkg *mkfile.Package
+
+	genCompileCommand bool
+	makeFileOnly      bool
+	cleanBuild        bool
+	output            string
+}
+
+func newBuildCommand() *cobra.Command {
+	c := &buildCommand{}
+	cmd := &cobra.Command{
+		Use:   "build",
+		Short: "build command builds a project",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if c.cleanBuild {
+				return c.clean()
+			}
+			return c.build(args)
+		},
+	}
+	cmd.Flags().BoolVarP(&c.makeFileOnly, "makefile", "m", false, "generate makefile and exits")
+	cmd.Flags().BoolVarP(&c.cleanBuild, "clean", "c", false, "clean stage directory")
+	cmd.Flags().BoolVarP(&c.genCompileCommand, "compile_command", "p", false, "generate compile_commands.json for IDE")
+	cmd.Flags().StringVarP(&c.output, "output", "o", "", "output file name")
+	return cmd
+}
+
+func (c *buildCommand) parsePackage(root string) error {
+	abspath, err := filepath.Abs(root)
+	if err != nil {
+		return err
+	}
+
+	loader := mkfile.NewLoader()
+	pkg, err := loader.Load(abspath)
+	if err != nil {
+		return err
+	}
+
+	b := mkfile.NewBuilder().
+		WithCxxFlags(c.cxxFlags).
+		WithLDFlags(c.ldflags)
+
+	err = b.Parse(pkg)
+	if err != nil {
+		return err
+	}
+	c.builder = b
+	c.entryPkg = pkg
+	return nil
+}
+
+func (c *buildCommand) xchainRoot() (string, error) {
+	xroot := os.Getenv("XROOT")
```

Comment:
We have a XCHAIN_ROOT, is this the same with XCHAIN_ROOT?

Reply:
Yes, new code has reused XCHAIN_ROOT.